Interrogator.xUnit/NotConcurrentAttribute.cs

```diff
@@ -0,0 +1,189 @@
+			GroupName = $"{groupName}_{scope}"; // Adding scope to group name ensures that different scoped attributes don't collide
+			Scope = scope;
+		}
+
+		internal override Result<Option<MethodInfo>, string> TryGetMethod(Type containingType, MemberInfo member, MethodInfo[] testMethods)
+		{
+			var result = (Type ?? containingType)
+				.TryGetMethod(member.Name, ParameterTypes)
+				.Select(currentMethod => GetPreviousInChain(currentMethod, containingType, Scope, GroupName, testMethods));
+			return result;
+		}
+
+		private static Option<MethodInfo> GetPreviousInChain(MethodInfo currentMethod, Type type, ConcurrencyScope scope, string groupName, MethodInfo[] executingMethods)
+		{
+			var methodSet = new HashSet<MemberInfo>(executingMethods);
+			var methodsToExecute = GetSameGroupMethods(scope, type, groupName).Where(methodSet.Contains).ToArray();
+
+			return GetPreviousMethod(currentMethod, methodsToExecute);
+		}
+
+		private static MethodInfo[] GetSameGroupMethods(ConcurrencyScope scope, Type type, string groupName)
+		{
+			IEnumerable<Type> types;
+			switch (scope)
+			{
+				case ConcurrencyScope.Class:
+					types = new[] { type };
+					break;
+				case ConcurrencyScope.Assembly:
+					types = GetTypesInAssembly(type);
+					break;
+				case ConcurrencyScope.Namespace:
+					types = GetTypesInNamespace(type);
+					break;
+				case ConcurrencyScope.ClassHierarchy:
+					types = GetTypesInClassHierarchy(type);
+					break;
+				default:
+					types = Array.Empty<Type>();
+					break;
+			}
+
+			var list = new List<MethodInfo>();
+			foreach (var t in types)
+			{
+				var key = (t, groupName);
+				if (_methodInfoDictionary.ContainsKey(key))
+					list.AddRange(_methodInfoDictionary[key]);
+			}
+
+			return list.ToArray();
+		}
+
+		private static MethodInfo[] GetMethodsInTypes(params Type[] types)
+			=> types
+				.OrderTypes()
+				.SelectMany(GetMethodsInType)
+				.ToArray();
+
+		private static IEnumerable<MethodInfo> GetMethodsInType(Type type)
+			=> type
+				.GetMethods()
+				.Where(m => m.GetCustomAttributes(true).OfType<NotConcurrentAttribute>().Any())
+				.OrderBy(GetMethodKey);
+
+		private static string GetMethodKey(MethodInfo method)
+			=> $"{method.Name}-{String.Join("-", method.GetParameters().Select(p => p.ParameterType.FullName))}";
+
+		private static IEnumerable<Type> GetTypesInClassHierarchy(Type type)
+		{
+			var types = new HashSet<Type>(GetParentTypes(type)) { type };
+
+			foreach (var parentType in types.ToArray())
+			{
+				foreach (var nested in GetNestedTypes(parentType))
+					types.Add(nested);
+			}
+
+			return types;
+		}
+
+		private static IEnumerable<Type> GetParentTypes(Type type)
+		{
+			var nextParent = type?.DeclaringType;
+			while (nextParent != typeof(object) && nextParent != null)
+			{
+				yield return nextParent;
+				nextParent = nextParent.DeclaringType;
+			}
+		}
+
+		private static IEnumerable<Type> GetNestedTypes(Type type)
+		{
+			foreach (var t in type.GetNestedTypes())
```

Comment:
This function currently serves to purpose. However, if made cursive then it would.